File: app/src/main/java/com/ue/colorful/feature/material/PaletteFragment.java
package com.ue.colorful.feature.material;

import android.app.Fragment;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import com.ue.colorful.R;
import com.ue.colorful.model.PaletteColor;
import com.ue.colorful.model.PaletteColorSection;

public class PaletteFragment extends Fragment implements ColorCardRecyclerAdapter.ColorCardRecyclerAdapterCallback {

    public static final String ARG_COLOR_SECTION = "COLOR_SECTION";

    private static final String SECTION_KEY = "SECTION_KEY";

    private RecyclerView mListView;
    private ColorCardRecyclerAdapter mRecyclerAdapter;
    private PaletteColorSection mPaletteColorSection = null;

    protected ClipboardManager mClipboardManager;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        setHasOptionsMenu(true);
        View v = inflater.inflate(R.layout.fragment_color_palette, container, false);
        mListView = (RecyclerView) v;
        return mListView;
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.palette_color_menu, menu);
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        if (savedInstanceState != null) {
            mPaletteColorSection = savedInstanceState.getParcelable(SECTION_KEY);
        }
        if (mPaletteColorSection == null) {
            mPaletteColorSection = getArguments().getParcelable(ARG_COLOR_SECTION);
        }

        mRecyclerAdapter = new ColorCardRecyclerAdapter(getActivity(), mPaletteColorSection.getPaletteColorList(), this);
        mListView.setLayoutManager(new LinearLayoutManager(getActivity()));
        mListView.setAdapter(mRecyclerAdapter);
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        outState.putParcelable(SECTION_KEY, mPaletteColorSection);
        super.onSaveInstanceState(outState);
    }

    public void replaceColorCardList(PaletteColorSection paletteColorSection) {
        mPaletteColorSection = paletteColorSection;
        mRecyclerAdapter = new ColorCardRecyclerAdapter(getActivity(),
                mPaletteColorSection.getPaletteColorList(), this);
        mListView.setAdapter(mRecyclerAdapter);
    }

    public void scrollToTop() {
        mListView.smoothScrollToPosition(0);
    }

    @Override
    public void onCopyColorClicked(PaletteColor paletteColor) {
        if (mClipboardManager == null) {
            mClipboardManager = (ClipboardManager) getActivity().getSystemService(Context.CLIPBOARD_SERVICE);
        }
        final ClipData clip
                = ClipData.newPlainText(getString(R.string.color_clipboard, paletteColor.getColorSectionName(),
                paletteColor.getBaseName()), paletteColor.getHexString());
        mClipboardManager.setPrimaryClip(clip);

        Toast.makeText(getActivity(), getString(R.string.color_copied, paletteColor.getHexString()),
                Toast.LENGTH_SHORT).show();
    }
}

File: app/src/main/java/com/ue/colorful/feature/material/PaletteFragment.kt
package com.ue.colorful.feature.material

import android.app.Fragment
import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.view.*
import android.widget.Toast
import com.ue.colorful.R
import com.ue.colorful.model.PaletteColor
import com.ue.colorful.model.PaletteColorSection
import kotlinx.android.synthetic.main.fragment_color_palette.*

class PaletteFragment : Fragment(), ColorCardRecyclerAdapter.ColorCardRecyclerAdapterCallback {
    private var mRecyclerAdapter: ColorCardRecyclerAdapter? = null
    private var mPaletteColorSection: PaletteColorSection? = null

    protected var mClipboardManager: ClipboardManager? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle): View? {
        setHasOptionsMenu(true)
        return inflater.inflate(R.layout.fragment_color_palette, container, false)
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.palette_color_menu, menu)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        if (savedInstanceState != null) {
            mPaletteColorSection = savedInstanceState.getParcelable(SECTION_KEY)
        }
        if (mPaletteColorSection == null) {
            mPaletteColorSection = arguments.getParcelable(ARG_COLOR_SECTION)
        }

        mRecyclerAdapter = ColorCardRecyclerAdapter(activity, mPaletteColorSection!!.paletteColorList, this)
        paletteCardList.layoutManager = LinearLayoutManager(activity)
        paletteCardList.adapter = mRecyclerAdapter
    }

    override fun onSaveInstanceState(outState: Bundle) {
        outState.putParcelable(SECTION_KEY, mPaletteColorSection)
        super.onSaveInstanceState(outState)
    }

    fun replaceColorCardList(paletteColorSection: PaletteColorSection) {
        mPaletteColorSection = paletteColorSection
        mRecyclerAdapter = ColorCardRecyclerAdapter(activity,
                mPaletteColorSection!!.paletteColorList, this)
        paletteCardList?.adapter = mRecyclerAdapter
    }

    fun scrollToTop() {
        paletteCardList?.smoothScrollToPosition(0)
    }

    override fun onCopyColorClicked(paletteColor: PaletteColor) {
        if (mClipboardManager == null) {
            mClipboardManager = activity.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
        }
        val clip = ClipData.newPlainText(getString(R.string.color_clipboard, paletteColor.colorSectionName,
                paletteColor.baseName), paletteColor.hexString)
        mClipboardManager!!.primaryClip = clip

        Toast.makeText(activity, getString(R.string.color_copied, paletteColor.hexString),
                Toast.LENGTH_SHORT).show()
    }

    companion object {

        val ARG_COLOR_SECTION = "COLOR_SECTION"

        private val SECTION_KEY = "SECTION_KEY"
    }
}
